axelrod/strategies/memorytwo.py

```diff
@@ -1,11 +1,203 @@
+"""Memory Two strategies."""
+
+import itertools
+import warnings
+from typing import Tuple, Dict
+
 from axelrod.action import Action
 from axelrod.player import Player
+from axelrod.random_ import random_choice
 from .titfortat import TitForTat, TitFor2Tats
 from .defector import Defector
 
+
 C, D = Action.C, Action.D
 
 
+class MemoryTwoPlayer(Player):
+    """
+    Uses a sixteen-vector for strategies based on the last two rounds of play,
+    1.  P(C|CC, CC)
+    2.  P(C|CC, CD)
+    3.  P(C|CC, DC)
+    4.  P(C|CC, DD)
+    5.  P(C|CD, CC)
+    6.  P(C|CD, CD)
+    7.  P(C|CD, DC)
+    8.  P(C|CD, DD)
+    9.  P(C|DC, CC)
+    10. P(C|DC, CD)
+    11. P(C|DC, DC)
+    12. P(C|DC, DD)
+    13. P(C|DD, CC)
+    14. P(C|DD, CD)
+    15. P(C|DD, DC)
+    16. P(C|DD, DD))
+    Cooperator is set as the default player if sixteen_vector is not given.
+
+    Names
+
+    - Memory Two: [Hilbe2017]_
+    """
+
+    name = 'Generic Memory Two Player'
+    classifier = {
+        'memory_depth': 2,  # Memory-two Sixteen-Vector
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def __init__(self, sixteen_vector: Tuple[float, float, float, float,
+                                             float, float, float, float,
+                                             float, float, float, float,
+                                             float, float, float, float] = None,
+                 initial: Action = C) -> None:
+        """
+        Parameters
+        ----------
+
+        sixteen_vector: list or tuple of floats of length 16
+            The response probabilities to the preceding round of play
+        initial: C or D
+            The initial 2 moves
+        """
+        super().__init__()
+        self._initial = initial
+        self.set_initial_sixteen_vector(sixteen_vector)
+
+    def set_initial_sixteen_vector(self, sixteen_vector):
+        if sixteen_vector is None:
+            sixteen_vector = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
+            warnings.warn("Memory two player is set to default, Cooperator.")
+
+        self.set_sixteen_vector(sixteen_vector)
+        if self.name == 'Generic Memory Two Player':
+            self.name = "%s: %s" % (self.name, sixteen_vector)
+
+    def set_sixteen_vector(self, sixteen_vector: Tuple):
+        if not all(0 <= p <= 1 for p in sixteen_vector):
+            raise ValueError("An element in the probability vector, {}, is not "
+                             "between 0 and 1.".format(str(sixteen_vector)))
+
+        states = [(hist[:2], hist[2:])
+                  for hist in list(itertools.product((C, D), repeat=4))]
+
+        self._sixteen_vector = dict(zip(states, sixteen_vector)) # type: Dict[tuple, float]
```

Comment:
@marcharper this no longer maps `float` on to `states`, it's a change I suggested to @Nikoleta-v3 when trying to get the type hints to work https://github.com/Axelrod-Python/Axelrod/issues/8. 

I couldn't see why we needed to `map(float` but we do do it for `memory_one` so just highlighting in case this is something I've forgotten/missed (if indeed it's not needed, let's get rid of it from `memory_one` too).